cmake/README.md

```diff
@@ -1,23 +1,101 @@
 
-1. CMake in Cocos2d-x support platform iOS, Android, macOS, Linux, Win32.
+### Generate Visual Studio project
 
-1. Open your terminal, input `cmake --version`, if the cmake version is lower than 3.1, please upgrade
+```sh
+cd cocos2d-x
+mkdir win32-build && cd win32-build
+cmake .. -G"Visual Studio 15 2017" -Tv141
+```
+
+Execute `cmake --build .` to compile, or open `Cocos2d-x.sln` in Explorer to use the generated project. 
+
+### Generate macOS Project
+
+```sh
+cd cocos2d-x
+mkdir mac-build && cd msc-build
+cmake .. -GXcode
+open Cocos2d-x.xcodeproj
+```
+
+### Generate iOS Project
+
+```sh
+cd cocos2d-x
+mkdir ios-build && cd ios-build
+cmake .. -GXcode -DCMAKE_TOOLCHAIN_FILE=../cmake/ios.toolchain.cmake
+open Cocos2d-x.xcodeproj
+```
+
+### CMake in Android Studio
+
+We use the Gradle to Android applacition, and Gradle use cmake to build the native code, see the [property](https://github.com/cocos2d/cocos2d-x/blob/84be684e3858393a6f3efc50e3f95d4e0ac92a20/tests/cpp-empty-test/proj.android/gradle.properties#L38): `PROP_NDK_MODE`, it control the way of native build
+
+```sh
+# android cmake support
+# uncomment it, native code will build by cmake
+# keep comment, native code will build by ndkBuild
+PROP_NDK_MODE=cmake
+```
+
+If you want to add cmake build arguments, please add it at [externalNativeBuild](https://github.com/cocos2d/cocos2d-x/blob/84be684e3858393a6f3efc50e3f95d4e0ac92a20/tests/cpp-empty-test/proj.android/app/build.gradle#L25) block of [app/build.gradle] file.
+
+## Prebuilt feature
+
+To solve the problem that compiling engine sources spends too long time, we add the feature of pre-builds libs. Using this feature you only need build engine sources once for a specific environment.
+
+### Next is a example of build libs once, and use it in next build:
+
+```sh
+cocos new -l cpp -p my.pack.app1 test_app1
+mkdir app1_build && cd app1_build
+cmake ../test_app1 -DGEN_COCOS_PREBUILT=ON
+make prebuilt
+```
+
+Close option `GEN_COCOS_PREBUILT` and open `USE_COCOS_PREBUILT` to use prebuilt in the same project
+
+```sh
+cmake ../test_app1 -DGEN_COCOS_PREBUILT=OFF -DUSE_COCOS_PREBUILT=ON
+make TemplateCpp
+open bin/TemplateCpp.app
+```
+
+Add `-DUSE_COCOS_PREBUILT=ON` to use prebuilt libs in another cmake build.
+
+```sh
+cocos new -l cpp -p my.pack.app2 test_app2
+mkdir app2_build && cd app2_build
+cmake ../test_app2 -DUSE_COCOS_PREBUILT=ON
+make TemplateCpp
+open bin/TemplateCpp.app
+```
+> Any other cpp project can use prebuilt in this way
 
-2. You should use out-of-source build, that is said you need to create a different directory from `$cocos2dx_root` to execute cmake command.
+Using this feature on Android exists a little difference, for CMake can't find system environment when build in Gradle. So you need to [supply a path](https://github.com/cocos2d/cocos2d-x/blob/c087be314c2c56a757bf66163b173746b5d6ad34/tests/cpp-empty-test/proj.android/app/build.gradle#L34) as the location of prebuilt libs.
```

Comment:
It seems like there is access to env from gradle System.getenv("KEYSTORE")
https://stackoverflow.com/questions/25589052/proper-way-to-use-system-environment-variables-in-gradle-using-android-studio

Reply:
thanks, I will record this idea into https://github.com/cocos2d/cocos2d-x/issues/18717, might improve it in the future.